File: src/olcw78 (die yoong)/jsAlgorithmStudy/week02/10_괄호_회전하기.js
/**
 * 풀이
 *   (, {, [ 은 stack 에 넣고
 *   ), }, ] 은 stack pop 으로 비교 하여 쌍이 맞는지 체크
 * 분석
 *   TC: O(N^2)
 *   - 회전해주는 루프: O(N)
 *   - 내부 stack push/pop 하며 쌍이 맞는지 체크하는 루프: O(M)
 *   - O(N * M) -> O(N^2)
 *   SC: O(N)
 * @param s {string}
 * @return {number}
 * */
function solution(s) {
    if (s.length === 0) return 0

    const $s = s.split("")
    let rotCnt = 0
    let answer = 0
    while (rotCnt++ < s.length) {
        const joinedStr = $s.join("")
        if (isCorrect(joinedStr)) {
            answer++
        }
        $s.push($s.shift())
    }

    return answer
}

/**
 * @param str {string}
 * @return {boolean}
 */
function isCorrect(str) {
    const stack = []

    /** @type {(last: string, ch: string) => boolean} */
    const isLastAndCurrentMatch = (last, ch) => {
        const ok1 = ch === ")" && last === "("
        const ok2 = ch === "]" && last === "["
        const ok3 = ch === "}" && last === "{"

        return ok1 || ok2 || ok3
    }

    for (const ch of str) {
        if (ch === "[" || ch === "(" || ch === "{") {
            stack.push(ch)
        } else if (ch === "]" || ch === ")" || ch === "}") {
            const last = stack.pop()
            if (!isLastAndCurrentMatch(last, ch)) return false
        }
    }

    return 0 === stack.length
}

console.log(solution("[](){}"))
console.log(solution("}]()[{"))
console.log(solution("[)(]"))
console.log(solution("}}}"))
console.log(solution("(){{"))
console.log(solution("{(})"))
// 괄호를 열고 닫았을 때, 이 사이에 다른 괄호가 열려있다면 실패

File: src/olcw78 (die yoong)/jsAlgorithmStudy/week02/10_괄호_회전하기.go
package week02

import (
	"fmt"
	"strings"
)

func solution10(s string) int {
	if len(s) == 0 {
		return 0
	}

	str := strings.Split(s, "")
	rotCnt, answer := 0, 0

	for rotCnt < len(s) {
		joinedStr := strings.Join(str, "")
		if isCorrect(joinedStr) {
			answer++
		}

		first := str[0]
		str = append(str[1:], first)
		rotCnt++
	}

	return answer
}

func isCorrect(str string) bool {
	stack := make([]int32, 0, len(str)/2)
	isLastAndCurrentMatch := func(last, ch int32) bool {
		ok1 := ch == ']' && last == '['
		ok2 := ch == '}' && last == '{'
		ok3 := ch == ')' && last == '('
		return ok1 || ok2 || ok3
	}

	for _, ch := range str {
		if ch == '(' || ch == '{' || ch == '[' {
			stack = append(stack, ch)
		} else if ch == ')' || ch == '}' || ch == ']' {
			if len(stack) == 0 {
				return false
			}

			last := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if ok := isLastAndCurrentMatch(last, ch); !ok {
				return false
			}
		}
	}
	return 0 == len(stack)
}

func Run10() {
	r1 := solution10("(){}[]")
	fmt.Println(r1)

	r2 := solution10("}]()[{")
	fmt.Println(r2)

	r3 := solution10("[)(]")
	fmt.Println(r3)

	r4 := solution10("}}}")
	fmt.Println(r4)

	r5 := solution10("()}}")
	fmt.Println(r5)

	r6 := solution10("({)}")
	fmt.Println(r6)
}
